Add codecov.yml to fix path issue and update codecov uploader

harlantc opened the pull request:
supersedes #1321, done to fix head branch.

# Summary
Fix path parsing issue on codecov as well as updated for the new codecov uploader

# Details

In investigating the codecov reports jumping around a lot and just not making a whole lot of sense (#903), it seems the file tree for every report has been split between the normal project root, "sherpa" and a failed attempt by codecov to fix some more complex paths from conda builds to "work/sherpa/sherpa/sherpa", causing inconsistencies in the reports. trying to examine these files on codecov in the 'work/sherpa/sherpa/sherpa' dir reports as not found, leading to strange behavior. This means all our reports since the beginning is not accurate. 

This PR is to fix with codecov's [path fix](https://docs.codecov.com/docs/fixing-paths) and also introduce a way to configure reports for enhancements in the future if we want them (such as reorganizing reports by test type, system, etc, or changing how reports are evaluated and displayed).

This PR also fixes #1320, changing the codecov installation in CI from conda/pip (the deprecated bash version), instead, using the codecov actions from Github. This also plays into the previous issue surprisingly. The codecov Github action exposes the selection of reports, which revealed that reports returned to codecov was several files, all but one of them being coverage reports. These were (likely) interfering with codecov, resulting in the strange issues we've been seeing.

## specific changes
Add a codecov.yml
- sets path fix from "work/sherpa/sherpa" to "" to correct weird path settings
- sets the number of builds to 8, in order to match those returned from CI jobs.
- tell codecov to only report 'informative'ly, to keep reports from failing CI runs (TBD if this is kept)

Move codecov install for all CI tests into a codecov task
- This action uses the new uploader, solving the deprecation issue
- exposes the file selection for each report, setting it results in reports, not including unrelated files codecov guessed to be correct, possibly confusing reports.
- change coverage reports from "term" to "xml", codecov prefers xml anyway, and couldn't find them with term.

DougBurke commented:
@harlantc it looks like this needs to be rebased to pick up #1336 so that (hopefully) the tests will pass.

harlantc commented:
@DougBurke thanks for the advice, I've gone ahead and rebased/squished, and cleaned up the log some, I can do more if needed. I wanted to run 2 CI runs though to double-check my changes one final time now that #1336 is included.

DougBurke reviewed (approved):
LGTM, :shipit: 

Thanks for doing this. I have proven incapable of the patience needed to track down this sort of thing!

Marie-Terrell reviewed (approved):
Looks good! Great work! Thank you!

DougBurke commented:
@harlantc - just look at the smooth chart since your work landed - great work:

![screenshot](https://user-images.githubusercontent.com/224638/154319983-5a89bb27-f79b-486c-b613-eede327f9e6f.png)
